### PR TITLE
google-cloud-sdk: add support for components

### PR DESCRIPTION
**Pull request by xlambein**
###### Description of changes

This commit adds a set of Google Cloud SDK components in `pkgs.google-cloud-sdk.components` which can be installed along with the SDK through `pkgs.google-cloud-sdk.withExtraComponents [ ... ]`.

For example, the bigtable emulator can be installed with the following:

```bash
nix-build -E 'with import <nixpkgs> { }; google-cloud-sdk.withExtraComponents [google-cloud-sdk.components.bigtable]'
```

and later executed with:

```bash
result/bin/gcloud beta emulators bigtable start
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by xlambein** [comment]
Output from `nixpkgs-review`:

```
xlambein> nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"                                                                                                Mon 22 Aug 2022 10:40:47 AM CEST
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0
remote: Enumerating objects: 446, done.
remote: Counting objects: 100% (306/306), done.
remote: Compressing objects: 100% (108/108), done.
remote: Total 446 (delta 254), reused 224 (delta 198), pack-reused 140
Receiving objects: 100% (446/446), 704.97 KiB | 3.87 MiB/s, done.
Resolving deltas: 100% (267/267), completed with 113 local objects.
From https://github.com/NixOS/nixpkgs
   d6ef38857ed..eaf476b6080  master     -> refs/nixpkgs-review/0
$ git worktree add /home/xlambein/.cache/nixpkgs-review/rev-95a77857ef1db1825cbd02d038bacf02a3913111/nixpkgs eaf476b6080e0a7d60582cece4d6e4f6463b0814
Preparing worktree (detached HEAD eaf476b6080)
Updating files: 100% (31544/31544), done.
HEAD is now at eaf476b6080 Merge pull request #187080 from snpschaaf/add-mkdocs-swagger-ui-tag
$ nix-env --option system x86_64-linux -f /home/xlambein/.cache/nixpkgs-review/rev-95a77857ef1db1825cbd02d038bacf02a3913111/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit --no-ff 95a77857ef1db1825cbd02d038bacf02a3913111
Automatic merge went well; stopped before committing as requested
$ nix-env --option system x86_64-linux -f /home/xlambein/.cache/nixpkgs-review/rev-95a77857ef1db1825cbd02d038bacf02a3913111/nixpkgs -qaP --xml --out-path --show-trace --meta
Nothing to be built.
$ nix-shell /home/xlambein/.cache/nixpkgs-review/rev-95a77857ef1db1825cbd02d038bacf02a3913111/shell.nix
```

**Comment by xlambein** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/tools/admin/google-cloud-sdk/components.nix [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+  # Convert an archicecture + OS to a Nix platform
+  toNixPlatform = arch: os:
+    let
+      arch' = builtins.getAttr arch arches;
+      os' = builtins.getAttr os oses;
+    in
+    if (builtins.hasAttr arch arches && builtins.hasAttr os oses)
+    then "${arch'}-${os'}"
+    else throw "unsupported architecture '${arch}' and/or os '${os}'";
```
Comment (zimbatm) [inline pkgs/tools/admin/google-cloud-sdk/components.nix:33]:
```suggestion
    let
      arch' = arches.${arch} or (throw "unsupported architecture '${arch}')
      os' = oses.${os} or (throw "unsupported OS '${os}'");
    in
    "${os'}-${arch'}";
```
What do you think of this implementation?

File: pkgs/tools/admin/google-cloud-sdk/default.nix [hunk 2]
```diff
@@ -21,6 +21,12 @@ let
   sources = system:
     data.googleCloudSdkPkgs.${system} or (throw "Unsupported system: ${system}");
 
+  components = callPackage ./components.nix {
+    snapshotPath = builtins.fetchurl data.googleCloudSdkPkgs.components;
```
Comment (zimbatm) [inline pkgs/tools/admin/google-cloud-sdk/default.nix:25]:
```suggestion
    snapshotPath = builtins.fetchurl data.googleCloudSdkPkgs.components;
```
Move the components JSON into this repo. You'll find that Hydra doesn't allow to use `builtins.fetchurl` or IFD.

**Review by zimbatm — COMMENTED** [review]
nice, the overall design makes sense. The only issue is with the components.json that has to be pulled in this repo.

File: pkgs/tools/admin/google-cloud-sdk/components.nix (hunk 1 above)
Comment (xlambein) [reply pkgs/tools/admin/google-cloud-sdk/components.nix:33]: Much simpler, I like it a lot!

**Comment by zimbatm** [comment]
Result of `nixpkgs-review pr 187793` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by xlambein** [comment]
@zimbatm Changes done, thanks for the help :)

**Review by zimbatm — APPROVED** [review]
LGTM

**Pull request merged**